Refresh DEVELOPER.md (#561)
vertica/vertica-kubernetes: DEVELOPER.md

   ```shell
   make docker-build-operator docker-push-operator
   ```
2. Start the tests with the `run-int-tests` make target. These tests generate a large amount of output, so we recommend that you pipe the output to a file. This command uses `tee` to send output to STDOUT and a file named `kuttl.out`:
   ```shell
   make run-int-tests | tee kuttl.out
   ```

### Run Individual Tests

You can run individual tests from the command line with the [KUTTLE CLI](https://kuttl.dev/docs/cli.html#setup-the-kuttl-kubectl-plugin).

> **NOTE**
> Before you can run an individual test, you need to set up a communal endpoint. To set up a [MinIO](https://min.io/) endpoint, run the following make target:
>
> ```shell
> make setup-minio
> ```
>
> To set up a different communal endpoint, see [Custom communal endpoints](#custom-communal-endpoints).

To run an individual test, pass the `--test` command the name of a [test suite directory](./tests/). For example, this command runs all tests in the [http-custom-certs](./tests/e2e-leg-6/http-custom-certs/) directory:

```shell

> kubectl kuttl test --test http-custom-certs

Reviewer: Some late feedback about running a single test. There are some prereqs that need to be satisfied too with a `make init-e2e-env`. So maybe something like this:
```suggestion
make init-e2e-env && kubectl kuttl test --test http-custom-certs
```

```shell
[k8admin@docd01 vertica-kubernetes]$ docker image ls
REPOSITORY               TAG            IMAGE ID       CREATED          SIZE
...
quintush/helm-unittest   3.9.3-0.2.11   83c439b2cf46   9 months ago     105MB
...
```

Helm chart unit tests are stored in `helm-charts/verticadb-operator/tests` and use the [helm-unittest plugin](https://github.com/quintush/helm-unittest). The helm-unittest repo includes [test samples and templates](https://github.com/quintush/helm-unittest/tree/master/test/data/v3/basic) so you can model your own tests. For details about the test format, see the [helm-unittest GitHub repository](https://github.com/quintush/helm-unittest/blob/master/DOCUMENT.md).

Unit tests for the VerticaDB operator use the Go testing infrastructure. Some tests run the operator against a mock Kubernetes control plane created with [envtest](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/envtest). Per Go standards, test files are stored in package directories and end with `_test.go`.

## e2e Tests

> **IMPORTANT**
> The e2e tests only run on operators that were [deployed as an object](#deployment-object).

> The e2e tests use the [kuttl](https://github.com/kudobuilder/kuttl/) testing framework. Before you begin, set the default StorageClass for your Kubernetes cluster. Otherwise, most of the e2e tests use the `standard (default)` StorageClass. For details, see the [Kubernetes documentation](https://kubernetes.io/docs/tasks/administer-cluster/change-default-storage-class/).

Reviewer: The tests will use the default storageClass. The default can change, and in some environments a default isn't even set. If you are running with kind, a suitable default is picked and you don't even need to worry about this.

   ```
2. Start the tests with the `run-int-tests` make target. These tests generate a large amount of output, so we recommend that you pipe the output to a file. This command uses `tee` to send output to STDOUT and a file named `kuttl.out`:
   ```shell
   make run-int-tests | tee kuttl.out
   ```

### Run Individual Tests

You can run individual tests from the command line with the [KUTTLE CLI](https://kuttl.dev/docs/cli.html#setup-the-kuttl-kubectl-plugin).

> **NOTE**
> Before you can run an individual test, you need to set up a communal endpoint. To set up a [MinIO](https://min.io/) endpoint, run the following make target:
>
> ```shell
> make setup-minio
> ```
>
> To set up a different communal endpoint, see [Custom communal endpoints](#custom-communal-endpoints).

To run an individual test, pass the `--test` command the name of a [test suite directory](./tests/). For example, this command runs all tests in the [http-custom-certs](./tests/e2e-leg-6/http-custom-certs/) directory. The `--skip-delete` does not delete any resources that the tests create, which helps with debugging:

```shell
kubectl kuttl test --test http-custom-certs --skip-delete

> ```

Reviewer: You can remove mention of the `--skip-delete` option. We run with that by default now anyways.

| `PATH_PROTOCOL`         | Sets the Azure Block Storage protocol.                                                          | `azb://`         |
| `BUCKET_OR_CLUSTER`     | The account name.                                                                               | `account-name`   |
| `PATH_PREFIX`           | Identifies the developer that created the database. Must begin and end with a slash (`/`).      | `/username/`     |

### Pod logs

The e2e tests use [stern](https://github.com/stern/stern) to persist some pod logs to help debug any failures. The e2e tests create the `int-tests-output/` directory to store the logs.

The stern process completes when the kuttle tests run to completion. If you abort a kuttle test, then you must stop the stern process manually.

## Soak tests

The soak tests evaluate the operator over a long interval. The test is split into multiple iterations, and each iteration generates a random workload that is comprised of pod kills and scaling operations. If the tests succeed, the next iteration begins. You can set the number of iterations that the soak test runs.

> Soak tests are run with kuttl, and the random test generation is done with the [kuttle-step-gen tool](https://kuttl.dev/docs/testing/steps.html).

Reviewer: The kuttl-step-gen is an internal tool developed on this repo. Might make sense to not link to kuttl. Maybe cmd/kuttl-step-gen?
```suggestion
Soak tests are run with kuttl, and the random test generation is done with the [kuttl-step-gen tool](cmd/kuttl-step-gen/main.go).
```

To run the soak tests, create a configuration file that outlines the databases that you want to test and how you want the test framework to react. We provide a sample configuration file in [tests/soak/soak-sample.cfg](./tests/soak/soak-sample.cfg).

The following steps run the soak tests:

1. Create the databases that you want to test.
2. Copy the configuration file and make your edits:

   ```shell
   cp tests/soak/soak-sample.cfg local-soak.cfg
   vim local-soak.cfg
   ```

3. Set the number of iterations that you want to run with the `NUM_SOAK_ITERATIONS` environment variable:

   ```shell
   export NUM_SOAK_ITERATIONS=10
   ```

   To run infinite soak tests, set `NUM_SOAK_ITERATIONS` to `-1`.

4. To start the tests, run the following make target:

   ```shell
   make run-soak-tests
   ```

# Troubleshooting

The following sections provide troubleshooting tips for your deployment.

## Kubernetes Events

> The operator generates Kubernetes events for some key scenarios. This is helpful whne you need to understand what tasks the operator is working on. Use the following command to view the events:

Reviewer: ```suggestion
The operator generates Kubernetes events for some key scenarios. This is helpful when you need to understand what tasks the operator is working on. Use the following command to view the events:
```

  Normal  ClusterRestartSucceeded  28s    verticadb-operator  Successfully called 'admintools -t start_db' and it took 8.8401312s
```

## Retrieve `vertica.log`

You might need to inspect the contents of the `vertica.log` to diagnose a problem with the Vertica server. You can SSH into a container or use a sidecar logger.

> ### SSH into a container

Reviewer: It's technically not ssh. What about this?
```suggestion
### Exec into a container
```
